src/Contacts.ts

```diff
@@ -0,0 +1,209 @@
+import Client from './Client'
+import { privatePreferences } from '@xmtp/proto'
+import { EnvelopeWithMessage, buildUserPrivatePreferencesTopic } from './utils'
+
+export type ConsentState = 'allowed' | 'blocked' | 'unknown'
+
+export type ConsentListEntryType = 'address'
+
+export class ConsentListEntry {
+  value: string
+  entryType: ConsentListEntryType
+  permissionType: ConsentState
+
+  constructor(
+    value: string,
+    entryType: ConsentListEntryType,
+    permissionType: ConsentState
+  ) {
+    this.value = value
+    this.entryType = entryType
+    this.permissionType = permissionType
+  }
+
+  get key(): string {
+    return `${this.entryType}-${this.value}`
+  }
+
+  static fromAddress(
+    address: string,
+    permissionType: ConsentState = 'unknown'
+  ): ConsentListEntry {
+    return new ConsentListEntry(address, 'address', permissionType)
+  }
+}
+
+export class ConsentList {
+  entries: Map<string, ConsentState>
+  static _identifier: string
+
+  constructor() {
+    this.entries = new Map<string, ConsentState>()
+  }
+
+  allow(address: string) {
+    const entry = ConsentListEntry.fromAddress(address, 'allowed')
+    this.entries.set(entry.key, 'allowed')
+    return entry
+  }
+
+  block(address: string) {
+    const entry = ConsentListEntry.fromAddress(address, 'blocked')
+    this.entries.set(entry.key, 'blocked')
+    return entry
+  }
+
+  state(address: string) {
+    const entry = ConsentListEntry.fromAddress(address)
+    return this.entries.get(entry.key) ?? 'unknown'
+  }
+
+  static async getIdentifier(client: Client): Promise<string> {
+    if (!this._identifier) {
+      const { identifier } =
+        await client.keystore.getPrivatePreferencesTopicIdentifier()
+      this._identifier = identifier
+    }
+    return this._identifier
+  }
+
+  static async load(client: Client): Promise<ConsentList> {
+    const consentList = new ConsentList()
+    const identifier = await this.getIdentifier(client)
+    const contentTopic = buildUserPrivatePreferencesTopic(identifier)
+
+    const messages = await client.listEnvelopes(
+      contentTopic,
+      async ({ message }: EnvelopeWithMessage) => message
+    )
+
+    // decrypt messages
+    const { responses } = await client.keystore.selfDecrypt({
+      requests: messages.map((message) => ({ payload: message })),
+    })
+
+    // decoded actions
+    const actions = responses.reduce((result, response) => {
+      return response.result?.decrypted
+        ? result.concat(
+            privatePreferences.PrivatePreferencesAction.decode(
+              response.result.decrypted
+            )
+          )
+        : result
+    }, [] as privatePreferences.PrivatePreferencesAction[])
+
+    actions.forEach((action) => {
+      action.allow?.walletAddresses.forEach((address) => {
+        consentList.allow(address)
+      })
+      action.block?.walletAddresses.forEach((address) => {
+        consentList.block(address)
+      })
+    })
```

Comment:
I just discovered this in my RN testing but this will override for newer entries with older ones. You need to reverse the list before iterating.

Reply:
@neekolas to confirm, `client.listEnvelopes` should return these messages in ASC ordering, which is the correct order for processing?